### PR TITLE
Error handling in shell scripts

### PR DESCRIPTION
**Pull request by Unrud**
Most errors in the shell scripts for building, packaging and launching are ignored. (e.g. `make install` succeeds even when compiling fails).

This PR tries to mitigate the problem by always enabling `errexit`.

In bash scripts `pipefail` is enabled. In plain sh scripts I removed pipes or checked the output.

I removed some of the subshells used in program arguments, as errors are ignored even with `errexit`.

Besides the main focus of this PR, I added some missing quotes around path variables and replaced an instance of`*` with `./*` for globbing (preventing paths from being interpreted as options).

### CODE REVIEW

File: launch-game.sh [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
-if command -v mono >/dev/null 2>&1 && [ "$(grep -c .NETCoreApp,Version= ${ENGINEDIR}/bin/OpenRA.dll)" = "0" ]; then
+if command -v mono >/dev/null 2>&1 && [ "$(grep -c .NETCoreApp,Version= "${ENGINEDIR}/bin/OpenRA.dll")" = "0" ]; then
 	RUNTIME_LAUNCHER="mono --debug"
 else
 	RUNTIME_LAUNCHER="dotnet"
 fi
 
 if command -v python3 >/dev/null 2>&1; then
-	 LAUNCHPATH=$(python3 -c "import os; print(os.path.realpath('$0'))")
+	 LAUNCHPATH=$(python3 -c 'import os,sys; print(os.path.realpath(sys.argv[1]))' "$0")
```
Comment (anvilvapre) [inline launch-game.sh:12]: is/was LAUNCHPATH ever correct as it seems to include the basename `launch-game.sh`? or was it meant to be a file and not a dir.

alternative: `LAUNCHPATH=$(dirname $(reallink -f "$0"))` or `LAUNCHPATH=$(reallink -f "$0")`

with no python dependency.

File: launch-game.sh [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
 else
-	 LAUNCHPATH=$(python -c "import os; print(os.path.realpath('$0'))")
+	 LAUNCHPATH=$(python -c 'import os,sys; print(os.path.realpath(sys.argv[1]))' "$0")
 fi
```
Comment (anvilvapre) [inline launch-game.sh:15]: Similar.

File: launch-game.sh (hunk 1 above)
Comment (anvilvapre) [reply launch-game.sh:12]: bash -x on old script indeed shows `LAUNCHPATH=/home/../OpenRA/launch-game.sh`

File: launch-game.sh [hunk 3]
```diff
@@ -32,10 +34,10 @@ then
...
 
 # Launch the engine with the appropriate arguments
-${RUNTIME_LAUNCHER} ${ENGINEDIR}/bin/OpenRA.dll Engine.EngineDir=".." Engine.LaunchPath="${LAUNCHPATH}" ${MODARG} "$@"
+${RUNTIME_LAUNCHER} "${ENGINEDIR}/bin/OpenRA.dll" Engine.EngineDir=".." Engine.LaunchPath="${LAUNCHPATH}" ${MODARG} "$@" && rc=0 || rc=$?
 
 # Show a crash dialog if something went wrong
-if [ $? != 0 ] && [ $? != 1 ]; then
+if [ $rc != 0 ] && [ $rc != 1 ]; then
```
Comment (anvilvapre) [inline launch-game.sh]: trivial: "${rc}" like evereywhere

File: packaging/functions.sh [hunk 4]
```diff
@@ -140,8 +144,9 @@ install_data() {
 # Used by:
 #   Windows packaging
 #   Mod SDK Windows packaging
-install_windows_launcher()
-{
+install_windows_launcher() (
+	set -o errexit || exit $?
+
 	SRC_PATH="${1}"
```
Comment (anvilvapre) [inline packaging/functions.sh:150]: optional: sanity check to see if called from correct dir.

File: packaging/functions.sh [hunk 5]
```diff
@@ -150,14 +155,14 @@ install_windows_launcher()
...
 	FAQ_URL="${7}"
 
-	rm -rf "${SRC_PATH}/OpenRA.WindowsLauncher/obj"
+	rm -rf "${SRC_PATH}/OpenRA.WindowsLauncher/obj" || :
 	dotnet publish "${SRC_PATH}/OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj" -c Release -r "${TARGETPLATFORM}" -p:LauncherName="${LAUNCHER_NAME}" -p:TargetPlatform="${TARGETPLATFORM}" -p:ModID="${MOD_ID}" -p:DisplayName="${MOD_NAME}" -p:FaqUrl="${FAQ_URL}" -o "${DEST_PATH}" --self-contained true
 
 	# NET 6 is unable to customize the application host for windows when compiling from Linux,
 	# so we must patch the properties we need in the PE header.
 	# Setting the application icon requires an external tool, so is left to the calling code
 	python3 "${SRC_PATH}/packaging/windows/fixlauncher.py" "${DEST_PATH}/${LAUNCHER_NAME}.exe"
```
Comment (anvilvapre) [inline packaging/functions.sh:164]: trivial: check if dotnet and python3 are present

File: packaging/functions.sh [hunk 6]
```diff
@@ -263,14 +274,14 @@ install_linux_shortcuts() {
...
-			sed "s/{MODID}/${MOD_ID}/g" "${SRC_PATH}/packaging/linux/openra-mimeinfo.xml.in" | sed "s/{TAG}/${VERSION}/g" > "${SRC_PATH}/packaging/linux/openra-${MOD_ID}.xml"
+			sed -e "s/{MODID}/${MOD_ID}/g" -e "s/{TAG}/${VERSION}/g" "${SRC_PATH}/packaging/linux/openra-mimeinfo.xml.in" > "${SRC_PATH}/packaging/linux/openra-${MOD_ID}.xml"
 			install -m644 "${SRC_PATH}/packaging/linux/openra-${MOD_ID}.xml" "${BUILD_PATH}${SHARE_PATH}/mime/packages/openra-${MOD_ID}.xml"
 			rm "${SRC_PATH}/packaging/linux/openra-${MOD_ID}.xml"
 		fi
 
 		shift
 	done
-}
+)
```
Comment (anvilvapre) [inline packaging/functions.sh:284]: how could this have been missing. or did you add the start as well?

File: packaging/linux/AppRun.in [hunk 7]
```diff
@@ -1,6 +1,8 @@
 #!/bin/sh
+set -o errexit || exit $?
 
-HERE="$(dirname "$(readlink -f "${0}")")"
+LAUNCHER=$(readlink -f "${0}")
+HERE=$(dirname "$LAUNCHER")
 
 # Run the game or server
 if [ -n "$1" ] && [ "$1" = "--server" ]; then
```
Comment (anvilvapre) [inline packaging/linux/AppRun.in:8]: trivial: redundant check to see if not empty.

File: packaging/linux/buildpackage.sh [hunk 8]
```diff
@@ -44,9 +45,9 @@ fi
 # Add native libraries
 echo "Downloading appimagetool"
 if command -v curl >/dev/null 2>&1; then
-	curl -s -L -O https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage || exit 3
+	curl -s -L -O https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage
```
Comment (anvilvapre) [inline packaging/linux/buildpackage.sh:49]: quoted "http:// ...."

File: packaging/linux/buildpackage.sh [hunk 9]
```diff
@@ -44,9 +45,9 @@ fi
...
 else
-	wget -cq https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage || exit 3
+	wget -cq https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage
```
Comment (anvilvapre) [inline packaging/linux/buildpackage.sh:51]: trivial: set url once above

File: packaging/linux/openra.appimage.in [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
-cd "${HERE}/../lib/openra" || exit 1
+cd "${HERE}/../lib/openra"
 
 # APPIMAGE is an environment variable set by the runtime
 # defining the absolute path to the .AppImage file
 if [ -n "${APPIMAGE}" ]; then
 	LAUNCHER=${APPIMAGE}
 
 	APPIMAGEID=$(printf "file://%s" "${APPIMAGE}" | md5sum | cut -d' ' -f1)
+	test -n "${APPIMAGEID}"
```
Comment (anvilvapre) [inline packaging/linux/openra.appimage.in:14]: trivial: move to line 12

File: packaging/linux/openra.appimage.in [hunk 11]
```diff
@@ -43,21 +46,23 @@ fi
 
 # Run the game
 export SDL_VIDEO_X11_WMCLASS="openra-{MODID}-{TAG}"
-./OpenRA Game.Mod={MODID} Engine.LaunchPath="${LAUNCHER}" "${JOIN_SERVER}" "$@"
+./OpenRA Game.Mod={MODID} Engine.LaunchPath="${LAUNCHER}" "${JOIN_SERVER}" "$@" && rc=0 || rc=$?
 
 # Show a crash dialog if something went wrong
-if [ $? != 0 ] && [ $? != 1 ]; then
+if [ $rc != 0 ] && [ $rc != 1 ]; then
 	LOGS="${XDG_CONFIG_HOME:-${HOME}/.config}/openra/Logs"
```
Comment (anvilvapre) [inline packaging/linux/openra.appimage.in:53]: side note: storing logs and maps in `.config` better would be to use a user data dir,

File: packaging/linux/openra.appimage.in [hunk 12]
```diff
@@ -43,21 +46,23 @@ fi
...
 	if [ ! -d "${LOGS}" ] && [ -d "${HOME}/.openra/Logs" ]; then
 		LOGS="${HOME}/.openra/Logs"
 	fi
 
-	test -d Support/Logs && LOGS="${PWD}/Support/Logs"
+	if [ -d Support/Logs ]; then
```
Comment (anvilvapre) [inline packaging/linux/openra.appimage.in:58]: trivial: executable .. `-x` over `-d`

also: `"./Support/Logs"`.

is this a way to override the log location? for developers, or on macos?

File: packaging/linux/openra.in [hunk 13]
```diff
@@ -9,21 +11,23 @@ if [ "${1#${PROTOCOL_PREFIX}}" != "${1}" ]; then
...
 # Show a crash dialog if something went wrong
-if [ $? != 0 ] && [ $? != 1 ]; then
+if [ $rc != 0 ] && [ $rc != 1 ]; then
 	LOGS="${XDG_CONFIG_HOME:-${HOME}/.config}/openra/Logs"
 	if [ ! -d "${LOGS}" ] && [ -d "${HOME}/.openra/Logs" ]; then
 		LOGS="${HOME}/.openra/Logs"
 	fi
 
-	test -d Support/Logs && LOGS="${PWD}/Support/Logs"
+	if [ -d Support/Logs ]; then
```
Comment (anvilvapre) [inline packaging/linux/openra.in:23]: ` "./Support/Logs" `

File: packaging/macos/buildpackage.sh [hunk 14]
```diff
@@ -28,7 +29,7 @@ if [[ "${OSTYPE}" != "darwin"* ]]; then
 fi
 
 # Set the working dir to the location of this script
-cd "$(dirname "${0}")" || exit 1
+cd "$(dirname "${0}")"
```
Comment (anvilvapre) [inline packaging/macos/buildpackage.sh]: trivial: `pushd`

File: packaging/package-all.sh [hunk 15]
```diff
@@ -10,7 +12,7 @@ export GIT_TAG="$1"
 export BUILD_OUTPUT_DIR="$2"
```
Comment (anvilvapre) [inline packaging/package-all.sh:13]: check if dir exists.

File: packaging/windows/buildpackage.sh [hunk 16]
```diff
@@ -34,9 +34,9 @@ elif [[ ${TAG} == playtest* ]]; then
 fi
 
 if command -v curl >/dev/null 2>&1; then
-	curl -s -L -O https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe || exit 3
+	curl -s -L -O https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe
```
Comment (anvilvapre) [inline packaging/windows/buildpackage.sh:38]: `"http:..."`

File: packaging/windows/buildpackage.sh [hunk 17]
```diff
@@ -34,9 +34,9 @@ elif [[ ${TAG} == playtest* ]]; then
...
 else
-	wget -cq https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe || exit 3
+	wget -cq https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe
```
Comment (anvilvapre) [inline packaging/windows/buildpackage.sh:40]: `"https:.."`

**Review by anvilvapre — DISMISSED** [review]
please consider proposed changes.

File: packaging/linux/openra.appimage.in (hunk 11 above)
Comment (abcdefg30) [reply packaging/linux/openra.appimage.in:53]: See #10374. We should not change that in this PR.

File: launch-game.sh (hunk 3 above)
Comment (Unrud) [reply launch-game.sh]: Fixed

File: packaging/functions.sh (hunk 6 above)
Comment (Unrud) [reply packaging/functions.sh:284]: I modified the start as well.

File: packaging/functions.sh (hunk 5 above)
Comment (Unrud) [reply packaging/functions.sh:164]: Why? It just fails if they are not available.

**Comment by Unrud** [comment]
> please consider proposed changes.

I've changed variables from `$abc` to `${abc}` and changed the remaining occurrences of `cd "$(dirname "$0")"` to 
```
HERE=$(dirname "$0")
cd "${HERE}"
```

I agree with the rest of the suggestions, which I have not commented on, but I think they are beyond the scope of this PR. I would rather create separate PRs as I have found other problems with the scripts that I would like to fix.

File: packaging/linux/openra.appimage.in (hunk 12 above)
Comment (Unrud) [reply packaging/linux/openra.appimage.in:58]: > executable .. `-x` over `-d`

But it has to be a directory, right? What would be the point of this?

Comment (anvilvapre) [reply packaging/linux/openra.appimage.in:58]: > > executable .. `-x` over `-d`
> 
> But it has to be a directory, right? What would be the point of this?

trivial. even it its a directory it won't do you much good if you do not have executable rights on it.

**Comment by anvilvapre** [comment]
> > please consider proposed changes.
> 
> I've changed variables from `$abc` to `${abc}` and changed the remaining occurrences of `cd "$(dirname "$0")"` to
> 
> ```
> HERE=$(dirname "$0")
> cd "${HERE}"
> ```
> 
> I agree with the rest of the suggestions, which I have not commented on, but I think they are beyond the scope of this PR. I would rather create separate PRs as I have found other problems with the scripts that I would like to fix.

sure.

File: packaging/macos/buildpackage.sh (hunk 14 above)
Comment (Unrud) [reply packaging/macos/buildpackage.sh]: shellcheck says that pushd is undefined in POSIX sh

File: packaging/linux/openra.appimage.in (hunk 12 above)
Comment (Unrud) [reply packaging/linux/openra.appimage.in:58]: > even it its a directory it won't do you much good if you do not have executable rights on it.

It doesn't seem meaningful. Using `-d` is good enough. Checking with `-x` still doesn't mean that you can actually create (or write existing) log files and it has the additional caveat that it succeeds for executable files.

File: launch-game.sh (hunk 1 above)
Comment (pchote) [reply launch-game.sh:12]: `LAUNCHPATH` is supposed to contain the `launch-game.sh`. Also, `readlink -f` is a non-portable GNU-ism!

File: launch-dedicated.sh [hunk 18]
```diff
@@ -7,7 +7,7 @@
 #  Read the file to see which settings you can override
 
 ENGINEDIR=$(dirname "$0")
-if command -v mono >/dev/null 2>&1 && [ "$(grep -c .NETCoreApp,Version= ${ENGINEDIR}/bin/OpenRA.Server.dll)" = "0" ]; then
+if command -v mono >/dev/null 2>&1 && [ "$(grep -c .NETCoreApp,Version= "${ENGINEDIR}/bin/OpenRA.Server.dll")" = "0" ]; then
```
Comment (pchote) [inline launch-dedicated.sh:12]: The outer grep command is already double-quoted. Will this work as expected?

File: packaging/package-all.sh [hunk 19]
```diff
@@ -19,14 +21,12 @@ function build_package() (
 	}
 	#trap function executes on any error in the following commands
 	trap "on_build $1" ERR
-	set -e
 	echo "Building $1 package(s)."
 	cd "$1"
 	./buildpackage.sh "${GIT_TAG}" "${BUILD_OUTPUT_DIR}"
 )
 
 #exit on any non-zero exited (failed) command
```
Comment (pchote) [inline packaging/package-all.sh]: Should this comment be removed too?

Comment (Unrud) [reply packaging/package-all.sh]: Removed it.

File: launch-dedicated.sh (hunk 18 above)
Comment (Unrud) [reply launch-dedicated.sh:12]: Yes, nested quotes are already used in parameter expansions:

https://github.com/OpenRA/OpenRA/blob/c15af9f68abe070cf825fa1dcbf19921b8e8ce1d/launch-dedicated.sh#L16-L16

Comment (anvilvapre) [reply launch-dedicated.sh:12]: Nesting does not work I suspect. You concat 3 strings here. `${Name:-`, `Dedicated Server` and `}`. I.e. if you add `\? ` in there bash will expand it to a questionmark.

**Review by pchote — DISMISSED** [review]

**Comment by anvilvapre** [comment]
b.t.w. on my platform gtk-dialog does not work with `python` (2.7) but does with `python3`. likely because one of the imports fails (Gtk?).

not sure if it's worth checking to see if `python3` exists.

**Comment by Unrud** [comment]
> not sure if it's worth checking to see if `python3` exists.

Right. I think it should just default to **python3**. **python2** is end-of-life.

**Comment by Unrud** [comment]
I've added a commit to change it to **python3** in https://github.com/OpenRA/OpenRA/pull/20072.

**Comment by Unrud** [comment]
Rebased after merge of  #20072

**Review by anvilvapre — APPROVED** [review]
Reviewed in the past.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f063f326cf3c464bfe2fcff9c1ac8e683a73a0f5...935000e0f0c7d70aad4859f4d67fec90dab32651)

**Pull request merged**